presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TranslateExpressions.java

```diff
@@ -271,18 +294,250 @@ public Result apply(ValuesNode valuesNode, Captures captures, Context context)
+                    lambdaArgumentSymbolTypes.put(new Symbol(argument.getName().getValue()), type);
+                }
+                // the lambda expression itself
+                builder.put(NodeRef.of(lambdaExpression), functionType)
+                        // expressions from lambda arguments
+                        .putAll(lambdaArgumentExpressionTypes)
+                        // expressions from lambda body
+                        .putAll(getExpressionTypes(
+                                session,
+                                metadata,
+                                sqlParser,
+                                TypeProvider.copyOf(lambdaArgumentSymbolTypes),
+                                lambdaExpression.getBody(),
+                                emptyList(),
+                                NOOP));
+            }
+        }
+        for (RowExpression argument : aggregation.getArguments()) {
+            if (!isExpression(argument) || castToExpression(argument) instanceof LambdaExpression) {
+                continue;
+            }
+            builder.putAll(analyze(castToExpression(argument), session, typeProvider));
+        }
+        if (aggregation.getFilter().isPresent() && isExpression(aggregation.getFilter().get())) {
+            builder.putAll(analyze(castToExpression(aggregation.getFilter().get()), session, typeProvider));
+        }
+        return builder.build();
+    }
+
+    private Map<NodeRef<Expression>, Type> analyze(Expression expression, Session session, TypeProvider typeProvider)
+    {
+        return getExpressionTypes(
+                session,
                 metadata,
                 sqlParser,
-                context.getSymbolAllocator().getTypes(),
-                ImmutableList.of(expression),
-                ImmutableList.of(),
-                NOOP,
-                false);
+                typeProvider,
+                expression,
+                emptyList(),
+                NOOP);
+    }
+
+    private RowExpression toRowExpression(Expression expression, Rule.Context context)
+    {
+        return toRowExpression(expression, context.getSession(), analyze(expression, context.getSession(), context.getSymbolAllocator().getTypes()));
+    }
 
-        return SqlToRowExpressionTranslator.translate(expression, types, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), context.getSession(), false);
+    private RowExpression toRowExpression(Expression expression, Session session, Map<NodeRef<Expression>, Type> types)
+    {
+        return SqlToRowExpressionTranslator.translate(expression, types, ImmutableMap.of(), metadata.getFunctionManager(), metadata.getTypeManager(), session, false);
+    }
+
+    private RowExpression removeOriginalExpression(RowExpression rowExpression, Session session, Map<NodeRef<Expression>, Type> types)
+    {
+        if (isExpression(rowExpression)) {
```

Comment:
comment not addressed (https://github.com/prestodb/presto/pull/12710#discussion_r286713457)